### ISSUE TITLE
Column ordering - ASP.NET Core/EF core Scaffolding

### ISSUE THREAD

**Issue by KiranKumarH** [aspnet/Scaffolding#398]
Hi,
 
I’m following the below article to create a sample asp.net core project. 
[https://docs.microsoft.com/en-us/aspnet/core/data/ef-mvc/intro](url)
 
I’m trying to generate the controllers and views using the scaffolding option (Add->NEW Scaffolded Item->MVC5 Controller with views, using EF) available when I right click on the controllers directory.
  
I see the views generated is having the controls in the alphabetic order. I want this to in the same order of the properties in the model.

Model code
`namespace CoreProj.Models
{
    public class Student
    {
        public int ID { get; set; }
        public string LastName { get; set; }
        public string FirstMidName { get; set; }
        public DateTime EnrollmentDate { get; set; }
        public virtual ICollection<Enrollment> Enrollments { get; set; }
    }
}`
 
Generated view(Index.cshtml) having the controls created in alphabetic order.
 
`@Html.DisplayNameFor(model => model.LastName)
@Html.DisplayNameFor(model => model.FirstMidName)
@Html.DisplayNameFor(model => model.EnrollmentDate)
` 
Please let me know how can I keep the column order same as that we have in the model? 

Issue happens only on VS2015. This works perfectly fine on VS2013.

Thanks,
Kiran

**Comment by rowanmiller** [comment]
BTW for scaffolding you could look at using reflection order that the properties are defined on the CLR type... then filtered down to the properties that are part of the EF model. We don't do this in EF, because reflection isn't guaranteed to give you the same order they are defined in code. This has caused quite a number of bugs when a different CPU architecture results in a different property order... which in turn can have subtle and hard to debug side effects on how complex models are built. We still haven't completely solved these issues on EF6, which is why in EF Core we always do an explicit order.

This wouldn't impact scaffolding though, as it's a one time generation. Reflection order is usually close to the order from code.

**Comment by rowanmiller** [comment]
BTW - here is a related feature request in EF Core, where we plan to store column order as a separate concept than the order of properties in the model - https://github.com/aspnet/EntityFramework/issues/2272.

**Comment by prafullbhosale** [comment]
@KiranKumarH can you provide the version of `Microsoft.VisualStudio.Web.CodeGeneration.*` packages in your project? 

This should have been fixed in version `1.0.0-preview4-final` for VS2015.

**Comment by KiranKumarH** [comment]
Hi Prafull,

Thanks for the email!

I have updated the "Microsoft.VisualStudio.Web.CodeGeneration.*" package to "1.0.0-preview4-final " from "1.0.0-preview2-final".

After the nugget package update I see the below error while generating the scaffolding. let me know if I'm missing something.

![image](https://cloud.githubusercontent.com/assets/25523273/22725420/5e9c2f2a-edf3-11e6-8a71-aed56b1baa73.png)

Attaching the sample project for your reference. I'm tiring to generate the MVC5 Controller with views, using EF for the entity called "student" in the model directory.

[CoreProj.zip](https://github.com/aspnet/Scaffolding/files/759836/CoreProj.zip)

Thanks in advance.

Regards,
Kiran

**Comment by prafullbhosale** [comment]
@KiranKumarH,

The project.json is missing an entry in the `tools` section and also the version of scaffolding packages needs to be changed:

```json
{
   "dependencies" : {
      ...
      "Microsoft.VisualStudio.Web.CodeGenerators.Mvc": {
         "version": "1.0.0-preview4-final",
         "type": "build"
      },
     "Microsoft.VisualStudio.Web.CodeGeneration.Tools": {
         "version": "1.0.0-preview4-final",
         "type": "build"
      }
   },
   "tools" : {
      ...
      "Microsoft.VisualStudio.Web.CodeGeneration.Tools" : {
         "version": "1.0.0-preview4-final",
         "imports": [
            "dnxcore50",
            "portable-net45+win8"
         ]
      }
   }
}
```

Note that for project.json based projects, the package `Microsoft.VisualStudio.Web.CodeGeneration.Tools` needs to be present in both the `tools` and `dependencies` section and both the versions need to match.

**Comment by KiranKumarH** [comment]
Hi Prafull,

Thanks for the update!

I have added the suggested changes to the project.json and I see the below error now.
![image](https://cloud.githubusercontent.com/assets/25523273/22774263/6601ae96-eecc-11e6-818a-2f5b639a9ae8.png)

Could you please modify the below project and let me know what i'm missing?
[TestCore.zip](https://github.com/aspnet/Scaffolding/files/763197/TestCore.zip)

Thanks in advance.

Regards,
Kiran

**Comment by prafullbhosale** [comment]
@KiranKumarH,
I am assuming you are trying to use the `TestCore.Data.SchoolContext` class as your DbContext for this scaffolding.

You need to register this DbContext in the Startup.cs class in the `ConfigureServices` method.

Current method: 

```cs
        public void ConfigureServices(IServiceCollection services)
        {
            // Add framework services.
            services.AddDbContext<ApplicationDbContext>(options =>
                options.UseSqlServer(Configuration.GetConnectionString("DefaultConnection")));

            services.AddIdentity<ApplicationUser, IdentityRole>()
                .AddEntityFrameworkStores<ApplicationDbContext>()
                .AddDefaultTokenProviders();

            services.AddMvc();

            // Add application services.
            services.AddTransient<IEmailSender, AuthMessageSender>();
            services.AddTransient<ISmsSender, AuthMessageSender>();
        }
```
Add the below line in the `ConfigureServices` method
```cs
            services.AddDbContext<SchoolContext>(options => 
                options.UseSqlServer(Configuration.GetConnectionString("ConnectionStringName")));
```
and add an appropriate connection string in the `appsettings.json`

With the above change I was able to scaffold and controller with View using EF

```
dotnet aspnet-codegenerator -p project.json controller -m TestCore.Models.Course -dc TestCore.Data.SchoolContext -udl -f
Finding the generator 'controller'...
Running the generator 'controller'...
Attempting to compile the application in memory
Attempting to figure out the EntityFramework metadata for the model and DbContext: Course
info: Microsoft.Extensions.DependencyInjection.DataProtectionServices[0]
      User profile is available. Using 'C:\Users\prbhosal\AppData\Local\ASP.NET\DataProtection-Keys' as key repository and Windows DPAPI to encrypt keys at rest.
Added Controller : \.cs
Added View : \Views\Course\Create.cshtml
Added View : \Views\Course\Edit.cshtml
Added View : \Views\Course\Details.cshtml
Added View : \Views\Course\Delete.cshtml
Added View : \Views\Course\Index.cshtml
RunTime 00:00:09.73
```

Let me know if this helps.

**Comment by KiranKumarH** [comment]
@prafullbhosale 

I tried generating the views and controls with the suggested changes and I still see the contents of the view generated in the alphabetical order. I have tried generating the scaffolding using both command and visual studio.
_Ex:- Command_ : **dotnet aspnet-codegenerator -p project.json controller --force --controllerName StudentsController --model Core1.Models.Student --dataContext Core1.Data.SchoolContext --relativeFolderPath Controllers --controllerNamespace Core1.Controllers –-useDefaultLayout**  

Please refer to the below generated html code in alphabetical order in index.chtml view.

![image](https://cloud.githubusercontent.com/assets/25523273/22920079/3351e0ec-f2b9-11e6-8f93-2e2aea3bf2d6.png)

Where as I have the student model in the below order.

       public int ID { get; set; }
        public string LastName { get; set; }
        public string FirstMidName { get; set; }
        public DateTime EnrollmentDate { get; set; }
        public virtual ICollection<Enrollment> Enrollments { get; set; }

Basically, the controls generated in the views are in alphabetical order rather than the order in which properties are added in the student model.

Let me now if i'm missing some thing.

Thanks,
Kiran

**Comment by prafullbhosale** [comment]
@KiranKumarH, 
My previous [comment](https://github.com/aspnet/Scaffolding/issues/398#issuecomment-278092922)  was incorrect. 

This fix did not make it into the `1.0.0-preview4-final` version

It was fixed later for VS 2017 in `1.0.0-msbuild1-final` version.

Sorry about the confusion.

**Comment by KiranKumarH** [comment]
@prafullbhosale 

Thanks for the update.

is there a way to implement it in VS 2015? if yes, could you please modify the sample and share it with me.

Thanks,
Kiran

**Comment by prafullbhosale** [comment]
> I tried importing the package "1.0.0-msbuild1-final " and failed with the error. "Package Microsoft.VisualStudio.Web.CodeGeneration.Tools 1.0.0-msbuild1-final is not compatible with net462 (.NETFramework,Version=v4.6.2)"

This is expected. The packages with version `1.x-msbuild*` will **NOT** work with VS 2015 (or project.json based projects.)

> is there a way to implement it in VS 2015? if yes, could you please modify the sample and share it with me.

Modifying the project in this case won't help, as the scaffolders will internally sort them alphabetically anyways.

**Comment by zahik** [comment]
Added :
https://data.uservoice.com/forums/72025-entity-framework-feature-suggestions/suggestions/18936844-ef-core-migrations-column-ordering
